Modules/xSharePoint/DSCResources/MSFT_xSPWebAppPolicy/MSFT_xSPWebAppPolicy.psm1

```diff
@@ -56,63 +81,260 @@ function Set-TargetResource
-            }
-            "Full Control" {
-                $newRole = $wa.PolicyRoles.GetSpecialRole([Microsoft.SharePoint.Administration.SPPolicyRoleType]::FullControl)    
-            }
-            "Full Read" {
-                $newRole = $wa.PolicyRoles.GetSpecialRole([Microsoft.SharePoint.Administration.SPPolicyRoleType]::FullRead)    
-            }
+        $denyAll     = $wa.PolicyRoles.GetSpecialRole([Microsoft.SharePoint.Administration.SPPolicyRoleType]::DenyAll)
+        $denyWrite   = $wa.PolicyRoles.GetSpecialRole([Microsoft.SharePoint.Administration.SPPolicyRoleType]::DenyWrite)
+        $fullControl = $wa.PolicyRoles.GetSpecialRole([Microsoft.SharePoint.Administration.SPPolicyRoleType]::FullControl)
+        $fullRead    = $wa.PolicyRoles.GetSpecialRole([Microsoft.SharePoint.Administration.SPPolicyRoleType]::FullRead)
+
+        if (($SetCacheAccountsPolicy) -and (-not($wa.Properties.ContainsKey("portalsuperuseraccount")) -or -not($wa.Properties.ContainsKey("portalsuperreaderaccount")))) {
+            throw "Cache accounts not configured properly. PortalSuperUserAccount or PortalSuperReaderAccount properties is not configured."
         }
         
-        if ($wa.Policies.UserName -contains $params.UserName) {
-            $policyObject = $wa.Policies | Where-Object { $_.UserName -eq $params.UserName }
-        } else {
-            foreach($userName in $wa.Policies.UserName) {
-                $claimsPrincipal = New-SPClaimsPrincipal -EncodedClaim $userName -ErrorAction SilentlyContinue
-                if (($null -ne $claimsPrincipal) -and ($claimsPrincipal.Value -eq $params.UserName)) {
-                    $policyObject = $wa.Policies | Where-Object { $_.UserName -eq $userName }
+        $members = @()
+        foreach ($policy in $wa.Policies) {
+            $member = @{}
+            $member.Username = $policy.UserName
+            $member.PermissionLevel = $policy.PolicyRoleBindings.Name
+            $member.ActAsSystemAccount = $policy.IsSystemUser
+            $members += $member
+        }
+
+        if ($params.Members) {
+            Write-Verbose -Verbose "Processing Members parameter"
+
+            $allMembers = @()
+            foreach ($member in $params.Members) {
+                $allMembers += $member
+            }
+
+            if ($params.SetCacheAccountsPolicy) {
+                Write-Verbose -Verbose "Adding Cache Accounts to Members parameter"
+                $psuAccount = @{
+                    UserName = $wa.Properties["portalsuperuseraccount"]
+                    PermissionLevel = "Full Read"
+                }
+                $allMembers += $psuAccount
+                
+                $psrAccount = @{
+                    UserName = $wa.Properties["portalsuperreaderaccount"]
+                    PermissionLevel = "Full Read"
+                }
+                $allMembers += $psrAccount
+            }
```

Comment:
OK so basically if the web app is in claims mode we add them in claims format to web app policy and the property bag, if not then in default format. I'll make the updates to the logic here and push them back in when ready.